MathUtil: Simplify SaturatingCast implementation and fix edge case.

jordan-woyak opened the pull request:
Using `std::cmp_less` and `std::cmp_greater` handle the mismatched signedness logic for us.

Explicit `static_cast<T>` in the floating point path eliminated a warning produced in `MathUtilTest`:
```
MathUtil.h:68:17: warning: implicit conversion from 'const int' to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
```
Using `>=` instead of `>` fixed a floating point edge case that Dentomologist identified.

Dentomologist reviewed (changes requested):
If you add this test it fails:
```
  // Cast from a signed integer type to a smaller signed integer type
  EXPECT_EQ(-128, (MathUtil::SaturatingCast<s8, int>(-129)));
```

jordan-woyak commented:
> If you add this test it fails:
> 
> ```
>   // Cast from a signed integer type to a smaller signed integer type
>   EXPECT_EQ(-128, (MathUtil::SaturatingCast<s8, int>(-129)));
> ```

Oops, thanks for catching that. I've fixed this and added that test.

Dentomologist commented:
I found another annoying edge case, but this one is on master too.

This test fails with the call actually returning `-2147483648`:
```
EXPECT_EQ(2147483647, MathUtil::SaturatingCast<s32>(2147483648.0f));
```
Float values in the range `[2147483584, 2147483776]` are all equivalent to `2147483648` since floats can't represent every integer value. This causes `value > static_cast<T>(hi)` to return false for any `value` in that range because `hi` is also in that range, so they compare equal. It then goes to `return static_cast<Dest>(value);` which overflows to `-2147483648`.

This can be fixed by changing the condition to `value >= static_cast<T>(hi)`.
